src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -19,46 +18,86 @@ internal static class HPackEncoder
         //          without some additional guidance from the user about this.
         //          So for now, don't do dynamic encoding.
 
-        public static bool EncodeHeader(string name, string value, Span<byte> buffer, out int length)
+        public static byte[] EncodeHeader(string name, string value)
         {
-            int i = 0;
-            length = 0;
-
-            if (buffer.Length == 0)
+            byte[] fromPool = null;
+            Span<byte> span = stackalloc byte[256];
+            while (true)
             {
-                return false;
-            }
-
-            buffer[i++] = 0;
+                if (EncodeHeader(name, value, span, out int length))
+                {
+                    byte[] result = span.Slice(0, length).ToArray();
+                    if (fromPool != null)
+                    {
+                        ArrayPool<byte>.Shared.Return(fromPool);
+                    }
+                    return result;
+                }
 
-            if (i == buffer.Length)
-            {
-                return false;
+                int newLength = span.Length * 2;
```

Comment:
As it stands, this code will never be executed, because we don't have a known header long enough to exceed the stack buffer.

I'd suggest just asserting that the encode to the stack buffer succeeded. If we ever do add a large enough header, this assert will fail basically as soon as HttpClient is used, and we can revisit whether to increase the stack buffer size or do something else.

Reply:
(and similarly below in EncodeHeaderName)

Reply:
Ok.